[Jaegar Poon] iP

jaegarpoon opened the pull request:
### Ron Chatbot

> “Sometimes our stop-doing list needs to be bigger than our to-do list.” - Patti Digh ([source])(https://www.goodreads.com/work/quotes/13355889-four-word-self-help-simple-wisdom-for-complex-lives)

Ron is a chatbot personalised and improved from Duke with new built-in features for task tracking. 

Event types supported:

- Todo
- Deadline
- Event

Take these steps to run the app:

1. Download it from [here](https://github.com/jaegarpoon/ip/releases/tag/A-Jar).
2. Run it! 🙂

Best part - it is  **COMPLETELY FREE**!!

- [ ] CLI - easy to operate with just keyboard alone
- [ ] Saves your records locally and updates the task list each time you run it

Here's the main method:
`public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }`

Commit:
Add Gradle support

jaegarpoon committed:
Added Greet, Echo, Exit.

jaegarpoon committed:
Add, List.

jaegarpoon committed:
Mark as Done.

jaegarpoon committed:
ToDos, Events, Deadlines

jaegarpoon committed:
Automated Text UI Testing.

jaegarpoon committed:
Handle Errors.

jaegarpoon committed:
Delete.

jaegarpoon committed:
Save.

jaegarpoon committed:
Dates and Times.

jaegarpoon committed:
Use More OOP.

jaegarpoon committed:
Organize into packages

jaegarpoon committed:
Add JUnit tests

jaegarpoon committed:
Add JavaDoc comments

jaegarpoon committed:
Follow the Coding Standard

jaegarpoon committed:
Revert "Add JavaDoc comments"

This reverts commit 66151081b35e71a437d67ee33f116f067017ad85.

jaegarpoon committed:
Find

jaegarpoon committed:
Merge pull request #1 from jaegarpoon/branch-Level-9

Branch Level 9

jaegarpoon committed:
Add JavaDoc comments

jaegarpoon committed:
Merge pull request #2 from jaegarpoon/branch-A-JavaDoc

Add JavaDoc comments

jaegarpoon committed:
Merge pull request #3 from nus-cs2103-AY2122S2/add-gradle-support

Add Gradle support

jaegarpoon committed:
Use Gradle

jaegarpoon committed:
Merge pull request #4 from jaegarpoon/add-gradle-support

Use Gradle

jaegarpoon committed:
Update gradle support to fix errors

jaegarpoon committed:
Merge pull request #5 from jaegarpoon/add-gradle-support

Update gradle support to fix errors

jaegarpoon committed:
Use checkstyle

jaegarpoon committed:
Merge pull request #6 from jaegarpoon/branch-A-checkstyle

Use checkstyle

domlimm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+
+import duke.exception.RonException;
+
+/**
+ * Custom chatbot adapted from Duke
+ * Name of chatbot: Ron
+ *
+ * @author jaegarpoon
+ * @version 0.1
+ * @since 2022-02-21
+ */
```
Love this! 😄

domlimm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+ * @version 0.1
+ * @since 2022-02-21
+ */
+public class Duke {
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
+
+    public Duke(String filePath) {
+        this.ui = new Ui();
+        this.storage = new Storage(filePath);
+        try {
+            this.tasks = new TaskList(storage.load());
+        } catch (RonException e) {
+            this.tasks = new TaskList(filePath);
+            this.ui.showLoadingError();
+        }
+    }
```
Very succinct!

domlimm commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    public void printTasks() {
+        int i = 1;
```
Do you think it will be clearer if `i` is given a better name? Likewise in `findTasks`! Also, I really love your coding style 😍 

```suggestion
        int taskId = 1;
```

domlimm commented on src/main/java/duke/exception/RonException.java:
```diff
@@ -0,0 +1,11 @@
+public class RonException extends Exception {
+    @Override
+    public String toString() {
+        return "OH NO! ";
```
You could avoid magic literals as per [this](https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers).

```suggestion
        String generalMessage = "OH NO! ";
        return generalMessage;
```

domlimm commented on src/main/java/duke/parser/InputParser.java:
```diff
@@ -0,0 +1,113 @@
+        } else if (input.contains("deadline")) {
+            if (input.replace("deadline", "").trim().length() == 0) {
+                throw new EmptyDescriptionException("deadline");
+            } else if (trimmedText.charAt(trimmedText.length() - 1) == "/".charAt(0)) {
+                throw new MissingDateException();
+            } else if (!input.contains("/")) {
+                throw new WrongDateSyntaxException();
+            }
+            Task task = new Deadline(input);
+            if (task.getDescription().trim().length() == 0) {
+                throw new EmptyDescriptionException("deadline");
+            }
+            tasks.add(task);
+            System.out.println("Task added!");
+            System.out.println(task);
+            System.out.println("There are " + tasks.size() + " task(s) in the list.");
+        } else if (input.contains("event")) {
+            if (input.replace("event", "").trim().length() == 0) {
+                throw new EmptyDescriptionException("event");
+            } else if (trimmedText.charAt(trimmedText.length() - 1) == "/".charAt(0)) {
+                throw new MissingDateException();
+            } else if (!input.contains("/")) {
+                throw new WrongDateSyntaxException();
+            }
+            Task task = new Event(input);
+            tasks.add(task);
+            System.out.println("Task added!");
+            System.out.println(task);
+            System.out.println("There are " + tasks.size() + " task(s) in the list.");
+        } else if (input.contains("find")) {
+            if (input.replace("find", "").trim().length() == 0) {
+                throw new EmptyDescriptionException("find");
+            }
+            System.out.println("Here are some tasks that match your request:");
+            tasks.findTasks(input.substring(5));
+        } else {
+            throw (new UnidentifiedException());
+        }
+    }
+}
```
This method could be too lengthy. Do you think you could possibly make some modifications and shorten it?

Even though it's lengthy, but I understood everything written because of how you coded and named your variables. Great job! 😄

domlimm commented on src/main/java/duke/parser/StorageParser.java:
```diff
@@ -0,0 +1,46 @@
+
+        if (this.task instanceof Deadline) {
+            String dateTime = ((Deadline) this.task)
+                    .getDate().format(DateTimeFormatter.ofPattern("d/M/yyyy")) + " " + ((Deadline) this.task)
+                    .getTime().format(DateTimeFormatter.ofPattern("HHmm"));
```
Minor coding standard violation: you could bring the `+` down to Line 33.

```suggestion
                    .getDate().format(DateTimeFormatter.ofPattern("d/M/yyyy")) + " "
                    + ((Deadline) this.task).getTime().format(DateTimeFormatter.ofPattern("HHmm"));
```

domlimm reviewed (commented):
Overall, it was really pleasant to review your PR! I really enjoyed doing it. Everything was so neatly written. You followed the coding standard really closely, great job! Slight modifications to work on for the code quality, but it's near perfect! Good job Jaegar! 🍻

tiewweijian commented on src/main/java/duke/exception/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,21 @@
+ * Exception thrown when user does not input a description of task
+ */
+public class EmptyDescriptionException extends RonException {
+    public static final String MESSAGE = "Please add a description for ";
+    public static final String MESSAGE2 = " task.";
```
Perhaps the naming of these variables could be a bit more descriptive?

tiewweijian commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,36 @@
+package duke;
```
Great that you classified your packages differently, and similar for the other classes in `Duke.exception`, etc

tiewweijian commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Writes current state snapshot of tasklist into .txt file
+     *
+     * @throws WriteException
```
Perhaps you could add a one-liner statement on when will the exception be thrown.

tiewweijian commented on src/main/java/duke/exception/DeleteIndexException.java:
```diff
@@ -0,0 +1,14 @@
+public class DeleteIndexException extends RonException {
+    public static final String MESSAGE = "Please add index to delete.";
+
+    @Override
+    public String toString() {
+        return super.toString() + MESSAGE;
+    }
```
Great use on overriding `toString()` methods in all the classes!

tiewweijian reviewed (commented):
Overall great job, follows the Java coding standards, and not many violations. I enjoyed reading your codes, and your codes are easy to understand and follow. Keep up the good work!

jaegarpoon committed:
Add a GUI to Duke

jaegarpoon committed:
Merge pull request #7 from jaegarpoon/branch-Level-10

Add a GUI to Duke

jaegarpoon committed:
Include assertions to InputParser class

Assertions to check for non-empty task list on calling for input types
that require at least one task present (eg. mark, unmark, find, delete)

jaegarpoon committed:
Improve code quality

Following the coding standards provided from the textbook, several
changes are made to the code to ensure adherance. Changes made based on
reviews on GitHub as well.

jaegarpoon committed:
Use streams for selected loops

There are several nested for and if conditions within task list, which
are replaced by streams for ease of understanding.

jaegarpoon committed:
Merge pull request #8 from jaegarpoon/branch-A-CodeQuality

Branch a code quality

jaegarpoon committed:
Merge pull request #9 from jaegarpoon/branch-A-Streams

Use streams for selected loops

jaegarpoon committed:
Set up Continuous Integration

Workflow specified in gradle.yml

jaegarpoon committed:
Modify import statement order

jaegarpoon committed:
Update DukeTest for GitHub CI

jaegarpoon committed:
Deal with duplicate items

Adds the ability to recognize and deal with duplicate items.

jaegarpoon committed:
Merge pull request #10 from jaegarpoon/branch-C-DetectDuplicates

Deal with duplicate items

jaegarpoon committed:
Improve GUI

jaegarpoon committed:
Edit GUI commit

jaegarpoon committed:
User Guide and Ui screenshot

jaegarpoon committed:
Set theme jekyll-theme-slate

jaegarpoon committed:
Edit based on iP code quality

jaegarpoon committed:
Merge branch 'master' of https://github.com/jaegarpoon/ip

jaegarpoon committed:
Update for final submission
